libs/pdi-scanner/src/protocol/packets.rs

```diff
@@ -0,0 +1,1217 @@
+
+    /// This command turns-off the sense array light source. This is the DEFAULT
+    /// mode.
+    ///
+    /// `ASCII character 6 = (36H)`
+    ///
+    /// # Response
+    ///
+    /// No response.
+    TurnArrayLightSourceOffRequest,
+
+    /// This command causes the scanner’s motor to run in the forward direction
+    /// (ejecting a document from the rear of the unit). Motor runs until exit
+    /// sensors say that document has been ejected, or runs for a max run time
+    /// of about 4 seconds. ASCII character 3 = (33H)
+    ///
+    /// # Response
+    ///
+    /// No response.
+    EjectDocumentToRearOfScannerRequest,
+
+    /// This command causes the scanner to eject a form (after scanning) at the
+    /// front input throat of the scanner, but form remains gripped by the
+    /// scanner input rollers. ASCII character 1 = (31H)
+    ///
+    /// # Response
+    ///
+    /// No response.
+    EjectDocumentToFrontOfScannerAndHoldInInputRollersRequest,
+
+    /// This command causes the scanner’s motor to run in the reverse direction
+    /// (clearing a document from the front entrance of the scanner). Motor runs
+    /// until front sensors indicate form has exited, or for a max run time of
+    /// about 4 seconds. ASCII character 4 = (34H)
+    ///
+    /// # Response
+    ///
+    /// No response.
+    EjectDocumentToFrontOfScannerRequest,
+
+    /// This command causes the scanner to eject a document held in escrow (rear
+    /// rollers), by advancing the feed mechanism only enough to release the
+    /// document. ASCII character 7 = (37H)
+    ///
+    /// # Response
+    ///
+    /// No response.
+    EjectEscrowDocumentRequest,
+
+    /// On receipt of this command, the scanner will re-scan a document in
+    /// escrow position (held by rear set of rollers), and re-transmit the data.
+    /// ASCII character [ = (5BH)
+    ///
+    /// # Response
+    ///
+    /// No response.
+    RescanDocumentHeldInEscrowPositionRequest,
+
+    EnableDoubleFeedDetectionRequest,
+    DisableDoubleFeedDetectionRequest,
+    CalibrateDoubleFeedDetectionRequest(DoubleFeedDetectionCalibrationType),
+    SetDoubleFeedDetectionSensitivityRequest {
+        percentage: ClampedPercentage,
+    },
+    SetDoubleFeedDetectionMinimumDocumentLengthRequest {
+        length_in_hundredths_of_an_inch: u8,
+    },
+
+    /// Requests the double feed detection LED intensity (`n3a30`).
+    GetDoubleFeedDetectionLedIntensityRequest,
+
+    /// Requests the double feed detection calibration value for a single sheet
+    /// (`n3a30`).
+    GetDoubleFeedDetectionSingleSheetCalibrationValueRequest,
+
+    /// Requests the double feed detection calibration value for a double sheet
+    /// (`n3a20`).
+    GetDoubleFeedDetectionDoubleSheetCalibrationValueRequest,
+
+    /// Requests the double feed detection threshold value (`n3a10`).
+    GetDoubleFeedDetectionDoubleSheetThresholdValueRequest,
+
+    RawPacket(Vec<u8>),
+}
+
+macro_rules! checked {
```

Comment:
Just out of curiosity, why does this need to be a macro rather than a function?

Reply:
The answer to this is actually fairly instructive. You can write the function like so and it compiles fine as-is:

```rust
fn checked(command: Command, parser: fn(&[u8]) -> nom::IResult<&[u8], ()>) -> Vec<u8> {
    if let Ok(([], _)) = parser(&command.to_bytes()) {
        command.to_bytes()
    } else {
        panic!("Command {:?} failed to parse", command)
    }
}
```

However, as you start converting `checked!(…)` to `checked(…)` you notice that some of them become compile errors. This is because the type of `parser` doesn't match:

```
error[E0308]: mismatched types
   --> libs/pdi-scanner/src/protocol/packets.rs:622:17
    |
620 |             Self::SetSerialNumberRequest(serial_number) => checked(
    |                                                            ------- arguments to this function are incorrect
621 |                 Command::new(b"*").with_data(serial_number),
622 |                 parsers::set_serial_number_request,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected fn pointer, found fn item
    |
    = note: expected fn pointer `for<'a> fn(&'a _) -> std::result::Result<(&'a _, ()), nom::Err<nom::error::Error<&'a _>>>`
                  found fn item `for<'a> fn(&'a _) -> std::result::Result<(&'a _, &'a [u8; 8]), nom::Err<nom::error::Error<&'a _>>> {set_serial_number_request}`
note: function defined here
   --> libs/pdi-scanner/src/protocol/packets.rs:582:4
    |
582 | fn checked(command: Command, parser: fn(&[u8]) -> nom::IResult<&[u8], ()>) -> Vec<u8> {
    |    ^^^^^^^                   --------------------------------------------
```

You might then try to make `checked` be generic, but then you get this error:

```
error[E0308]: mismatched types
   --> libs/pdi-scanner/src/protocol/packets.rs:622:17
    |
620 |             Self::SetSerialNumberRequest(serial_number) => checked(
    |                                                            ------- arguments to this function are incorrect
621 |                 Command::new(b"*").with_data(serial_number),
622 |                 parsers::set_serial_number_request,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
    |
    = note: expected fn pointer `for<'a> fn(&'a _) -> std::result::Result<(&'a _, _), nom::Err<nom::error::Error<&'a _>>>`
                  found fn item `for<'a> fn(&'a _) -> std::result::Result<(&'a _, &'a [u8; 8]), nom::Err<nom::error::Error<&'a _>>> {set_serial_number_request}`
note: function defined here
   --> libs/pdi-scanner/src/protocol/packets.rs:582:4
    |
582 | fn checked<T>(command: Command, parser: fn(&[u8]) -> nom::IResult<&[u8], T>) -> Vec<u8> {
    |    ^^^^^^^                      -------------------------------------------
```

Perhaps there is a way to get past this, but I'm not sure what it is.

Reply:
Interesting! Thanks for walking me through it. Doesn't seem worth digging deeper to me